janitor/io.py

```diff
@@ -63,9 +65,45 @@ def read_csvs(
                     "Files cannot be concatenated"
                 )
         return pd.concat(
-            list(dfs_dict.values()),  # noqa: PD011
+            list(dfs_dict.values()),
             ignore_index=True,
-            sort=False,
+            sort=False,  # noqa: PD011
         )
     else:
         return dfs_dict
+
+
+def read_commandline(cmd: str, **kwargs) -> pd.DataFrame:
+    """
+    Read a CSV file based on a command-line command.
+
+    For example, you may wish to run the following command on `sep-quarter.csv`
+    before reading it into a pandas DataFrame:
+
+    ```bash
+    cat sep-quarter.csv | grep .SEA1AA
+    ```
+
+    In this case, you can use the following Python code to load the dataframe:
+
+    ```python
+    import janitor as jn
+    df = jn.io.read_commandline("cat data.csv | grep .SEA1AA")
+    ```
+
+    This function assumes that your command line command will return
+    an output that is parsable using pandas.read_csv and StringIO.
+    We default to using pd.read_csv underneath the hood.
+    Keyword arguments are passed through to read_csv.
+
+
+    :param cmd: Shell command to preprocess a file on disk.
+    :param kwargs: Keyword arguments that are passed through to pd.read_csv().
+    :returns: A pandas DataFrame parsed from the stdout of the underlying
+        shell.
+    """
+
+    check("cmd", cmd, [str])
+    outcome = subprocess.run(cmd, shell=True, capture_output=True, text=True)
+    outcome = outcome.stdout
```

Comment:
Just wanted to bring this to attention again for discussion: I would prefer if we could not proceed with `pd.read_csv` if the return_code is non-zero.

Basically, there are at least 2 "kinds" of errors that can happen here, either the user enters a malformed shell cmd or the shell cmd is valid but the data itself is problematic (whether it is empty, or missing a comma, etc.) for `pd.read_csv` to parse.
Example of 1: `jn.io.read_commandline("catt data.csv")`
Example of 2: `jn.io.read_commandline("touch data.csv && cat data.csv")`

I am of the opinion that these 2 types of errors should not be categorized together. In particular, the former will (currently) lead to a pandas EmptyDataError, which is not actually the crux of the problem. I would very much prefer to catch the non-zero code, and raise a `JanitorError` containing the shell stderr.

Reply:
@thatlittleboy if the data has a non-zero code, the std error is outputted (since `capture_output` is set to True). for the other case where the data itself is problematic, I'd like to think that `pd.read_csv` should output the error. 

I probably misunderstood your comment; maybe an example with the current code could be helpful.

Reply:
Sure, apologies for being unclear. Pertaining to error type1, at the moment:

```shell
In [13]: janitor.io.read_commandline("echoo 'a,b\n1,2'")
---------------------------------------------------------------------------
EmptyDataError                            Traceback (most recent call last)
<ipython-input-13-12d5880211fb> in <module>
----> 1 janitor.io.read_commandline("echoo 'a,b\n1,2'")

/workspaces/pyjanitor/janitor/io.py in read_commandline(cmd, **kwargs)
    107     outcome = subprocess.run(cmd, shell=True, capture_output=True, text=True)
    108     outcome = outcome.stdout
--> 109     return pd.read_csv(StringIO(outcome), **kwargs)

/opt/conda/lib/python3.7/site-packages/pandas/util/_decorators.py in wrapper(*args, **kwargs)
    309                     stacklevel=stacklevel,
    310                 )
--> 311             return func(*args, **kwargs)
    312 
    313         return wrapper

/opt/conda/lib/python3.7/site-packages/pandas/io/parsers/readers.py in read_csv(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, error_bad_lines, warn_bad_lines, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)
    584     kwds.update(kwds_defaults)
    585 
--> 586     return _read(filepath_or_buffer, kwds)
    587 
    588 

/opt/conda/lib/python3.7/site-packages/pandas/io/parsers/readers.py in _read(filepath_or_buffer, kwds)
    480 
    481     # Create the parser.
--> 482     parser = TextFileReader(filepath_or_buffer, **kwds)
    483 
    484     if chunksize or iterator:

/opt/conda/lib/python3.7/site-packages/pandas/io/parsers/readers.py in __init__(self, f, engine, **kwds)
    809             self.options["has_index_names"] = kwds["has_index_names"]
    810 
--> 811         self._engine = self._make_engine(self.engine)
    812 
    813     def close(self):

/opt/conda/lib/python3.7/site-packages/pandas/io/parsers/readers.py in _make_engine(self, engine)
   1038             )
   1039         # error: Too many arguments for "ParserBase"
-> 1040         return mapping[engine](self.f, **self.options)  # type: ignore[call-arg]
   1041 
   1042     def _failover_to_python(self):

/opt/conda/lib/python3.7/site-packages/pandas/io/parsers/c_parser_wrapper.py in __init__(self, src, **kwds)
     67         kwds["dtype"] = ensure_dtype_objs(kwds.get("dtype", None))
     68         try:
---> 69             self._reader = parsers.TextReader(self.handles.handle, **kwds)
     70         except Exception:
     71             self.handles.close()

/opt/conda/lib/python3.7/site-packages/pandas/_libs/parsers.pyx in pandas._libs.parsers.TextReader.__cinit__()

EmptyDataError: No columns to parse from file
```

so from the perspective of the end user, the problem is empty data, but the actual error is hidden away in the stderr

```shell
In [15]: subprocess.run("echoo 'a,b\n1,2'", shell=True, capture_output=True, text=True)
Out[15]: CompletedProcess(args="echoo 'a,b\n1,2'", returncode=127, stdout='', stderr='/bin/sh: 1: echoo: not found\n')
```

Unless I'm missing something, while stderr is captured with `capture_output`, it's not actually visible to the end user?

Reply:
totally clear now; @BaritoneBeard my apologies for misleading you on that. so @BaritoneBeard @thatlittleboy something like this: 
```py
 outcome = subprocess.run(cmd, shell=True, capture_output=True, text=True)
if outcome.returncode != 0:
    raise ValueError(outcome.stderr) # or JanitorError
else:
     outcome = outcome.stdout
```

?

Reply:
Ahhh, now I get it. @thatlittleboy, what you've suggested sounds like the saner thing to do. Thanks @samukweku for asking to clarify!

@BaritoneBeard, could you make this change, and do one final thing -- add this contribution to the CHANGELOG.md file please?